File: lib/test/t_GaussianProcess_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class Temporal Normal Process
 *
 *  Copyright 2005-2019 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
  {

    /* Default dimension parameter to evaluate the model */
    const UnsignedInteger defaultDimension = 1;

    /* Spatial dimension of the model */
    const UnsignedInteger inputDimension = 1;

    /* Amplitude values */
    Point amplitude(defaultDimension, 1.0);
    /* Scale values */
    Point scale(inputDimension, 1.0);

    const Scalar tmin = 0.0;
    const Scalar step = 0.1;
    const UnsignedInteger n = 11;

    RegularGrid myTimeGrid(tmin, step, n);
    const UnsignedInteger size = 25;


    /* Second order model with parameters */
    ExponentialModel myCovModel(scale, amplitude);
    fullprint << "myCovModel=" << myCovModel << std::endl;

    GaussianProcess myProcess1(myCovModel, myTimeGrid);
    fullprint << "myProcess1=" << myProcess1 << std::endl;
    fullprint << "is stationary? " << myProcess1.isStationary() << std::endl;
    myProcess1.setSamplingMethod(0);
    fullprint << "mean over " << size << " realizations=" << myProcess1.getSample(size).computeMean() << std::endl;
    myProcess1.setSamplingMethod(2);
    fullprint << "mean over " << size << " realizations=" << myProcess1.getSample(size).computeMean() << std::endl;

    /* With constant trend */
    TrendTransform trend(SymbolicFunction("t", "4.0"), myTimeGrid);
    GaussianProcess myProcess2(trend, myCovModel, myTimeGrid);
    fullprint << "myProcess2=" << myProcess2 << std::endl;
    fullprint << "is stationary? " << myProcess2.isStationary() << std::endl;
    myProcess1.setSamplingMethod(0);
    fullprint << "mean over " << size << " realizations=" << myProcess1.getSample(size).computeMean() << std::endl;
    myProcess1.setSamplingMethod(2);
    fullprint << "mean over " << size << " realizations=" << myProcess1.getSample(size).computeMean() << std::endl;

    /* With varying trend */
    TrendTransform trend3(SymbolicFunction("t", "sin(t)"), myTimeGrid);
    GaussianProcess myProcess3(trend3, myCovModel, myTimeGrid);
    fullprint << "myProcess3=" << myProcess3 << std::endl;
    fullprint << "is stationary? " << myProcess3.isStationary() << std::endl;
    myProcess1.setSamplingMethod(0);
    fullprint << "mean over " << size << " realizations=" << myProcess1.getSample(size).computeMean() << std::endl;
    myProcess1.setSamplingMethod(2);
    fullprint << "mean over " << size << " realizations=" << myProcess1.getSample(size).computeMean() << std::endl;
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}

File: python/test/t_GaussianProcess_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott

# Default dimension parameter to evaluate the model
inputDimension = 1
outputDimension = 1

# Amplitude values
amplitude = [1.0] * outputDimension
# Scale values
scale = [1.0] * inputDimension

tmin = 0.0
step = 0.1
n = 11

myTimeGrid = ot.RegularGrid(tmin, step, n)
size = 25

# Second order model with parameters
myCovModel = ot.ExponentialModel(scale, amplitude)
print("myCovModel = ", myCovModel)

myProcess1 = ot.GaussianProcess(myCovModel, myTimeGrid)
print("myProcess1 = ", myProcess1)
print("is stationary? ", myProcess1.isStationary())
myProcess1.setSamplingMethod(ot.GaussianProcess.CHOLESKY)
print("mean over ", size, " realizations = ", myProcess1.getSample(size).computeMean())
myProcess1.setSamplingMethod(ot.GaussianProcess.GALLIGAOGIBBS)
print("mean over ", size, " realizations = ", myProcess1.getSample(size).computeMean())

# With constant trend
trend = ot.TrendTransform(ot.SymbolicFunction("t", "4.0"), myTimeGrid)
myProcess2 = ot.GaussianProcess(trend, myCovModel, myTimeGrid)
myProcess2.setSamplingMethod(ot.GaussianProcess.GALLIGAOGIBBS)
print("myProcess2 = ", myProcess2)
print("is stationary? ", myProcess2.isStationary())
print("mean over ", size, " realizations= ", myProcess2.getSample(size).computeMean())

# With varying trend
trend3 = ot.TrendTransform(ot.SymbolicFunction("t", "sin(t)"), myTimeGrid)
myProcess3 = ot.GaussianProcess(trend3, myCovModel, myTimeGrid)
print("myProcess3 = ", myProcess3)
print("is stationary? ", myProcess3.isStationary())
myProcess3.setSamplingMethod(ot.GaussianProcess.CHOLESKY)
print("mean over ", size, " realizations = ", myProcess3.getSample(size).computeMean())
myProcess3.setSamplingMethod(ot.GaussianProcess.GALLIGAOGIBBS)
print("mean over ", size, " realizations = ", myProcess3.getSample(size).computeMean())

cov = ot.ExponentialModel([2.0] * inputDimension, [1.0, 2.0, 3.0])
model = ot.GaussianProcess(cov, myTimeGrid)
print("model=", model)
print("marginal=", model.getMarginal([0, 2]))

# FIX #2121
ot.RandomGenerator.SetSeed(0)
standard_deviation = 10.0
mesh = ot.Mesh([[0.0]])  # singleton
cov_matrix = ot.CovarianceMatrix(
    1, [standard_deviation**2]
)  # associated "covariance matrix"
covModel = ot.UserDefinedCovarianceModel(mesh, cov_matrix)

# Create the "Gaussian Process" discretized on a singleton
myProcess = ot.GaussianProcess(covModel, mesh)
myProcess.setSamplingMethod(ot.GaussianProcess.GALLIGAOGIBBS)
size = 1000
sample = ot.Sample(size, 1)
for i in range(size):
    sample[i] = myProcess.getRealization().getValues()[0]

ott.assert_almost_equal(sample.computeStandardDeviation()[0], 10.0676)

# Multivariate outputs
ot.RandomGenerator.SetSeed(1)
inputDimension = 1
outputDimension = 2

size = 1000
ot.ResourceMap.SetAsUnsignedInteger("GaussianProcess-GibbsMaximumIteration", 1000)
# Amplitude values
amplitude = [2.0, 3.0]
# Scale values
scale = [1.0] * inputDimension

myCovModel = ot.ExponentialModel(scale, amplitude)

myProcess4 = ot.GaussianProcess(myCovModel, myTimeGrid)
print("myProcess4 = ", myProcess4)
print("is stationary? ", myProcess4.isStationary())
myProcess4.setSamplingMethod(ot.GaussianProcess.CHOLESKY)
sample = myProcess4.getSample(size)
print("mean over ", size, " realizations = ", sample.computeMean())
print("variance over ", size, " realizations = ", sample.computeVariance())
myProcess4.setSamplingMethod(ot.GaussianProcess.GALLIGAOGIBBS)
sample = myProcess4.getSample(size)
print("mean over ", size, " realizations = ", sample.computeMean())
print("variance over ", size, " realizations = ", sample.computeVariance())

trend5 = ot.TrendTransform(
    ot.SymbolicFunction(["t"], ["sin(pi_ / 2 * t)", "2 * sin(pi_ / 2 * t)"]), myTimeGrid
)
myProcess5 = ot.GaussianProcess(trend5, myCovModel, myTimeGrid)
print("myProcess5 = ", myProcess5)
print("is stationary? ", myProcess5.isStationary())
myProcess5.setSamplingMethod(ot.GaussianProcess.CHOLESKY)
sample = myProcess5.getSample(size)
print("mean over ", size, " realizations = ", sample.computeMean())
print("variance over ", size, " realizations = ", sample.computeVariance())
myProcess5.setSamplingMethod(ot.GaussianProcess.GALLIGAOGIBBS)
sample = myProcess5.getSample(size)
print("mean over ", size, " realizations = ", sample.computeMean())
print("variance over ", size, " realizations = ", sample.computeVariance())
